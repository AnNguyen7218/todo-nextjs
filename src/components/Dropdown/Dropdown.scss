@import 'config.colors';
@import 'config.fonts';

// use !important to over default style of dropdown
.dropdown-container {
  border-top: unset !important;
  border-left: unset !important;
  border-right: unset !important;
  border-bottom: 0.1rem solid $purple !important;
  border-radius: 0 !important;
  outline: unset !important;
  box-shadow: none !important;
}

.single-select {
  color: $purple !important;
  margin: 0 !important;
}

.indicator-icon {
  width: 2rem;
  height: 2rem;
  color: $purple;
}

.option {
  color: $black !important;
  background-color: $white !important;

  &:hover {
    background-color: $light-gray !important;
  }

  &--selected {
    color: $purple !important;
  }
}

.option-text {
  @include body;
}

.menu-container {
  border: unset !important;
  box-shadow: 0px 4px 16px rgba(33, 33, 33, 0.08) !important;
  border-radius: 0 !important;
}

.value-container {
  @include body;
  padding: 0 !important;
}

.dropdown-indicator-container {
  padding: 0 !important;
}
