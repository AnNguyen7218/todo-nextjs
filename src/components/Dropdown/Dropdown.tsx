import React from 'react';
import Select, {
  ValueType,
  components,
  ControlProps,
  ActionMeta,
  SingleValueProps,
  IndicatorProps,
  OptionProps,
  MenuProps,
  ValueContainerProps,
} from 'react-select';
import cx from 'classnames';
import ArrowDown from 'assets/icons/ic-arrow-down.svg';
import css from './Dropdown.scss';

// @NOTE dude to some scss import problem,
// in order to override class name by custom class name
// in your scss file please add 'div' string before your class name
// exp in scss file: div.my-custom-class {}

type Option = {
  value: string | number | boolean;
  label: string | JSX.Element;
};

type Options = Option[];

type DropdownProps = {
  value: Option;
  options: Options;
  onChange: (value: ValueType<Option>, actionMeta?: ActionMeta) => void;
  instanceId?: string;
  controlClassName?: string;
  singleValueClassName?: string;
  dropdownIndicatorClassName?: string;
  optionClassName?: string;
  optionSelectedClassName?: string;
  optionTextClassName?: string;
  menuClassName?: string;
  valueContainerClassName?: string;
};

type ControlCustomProps = ControlProps<Option> & {
  controlClassName?: string;
};

const Control: React.FC<ControlCustomProps> = ({
  controlClassName,
  ...props
}) => {
  return (
    <components.Control
      {...props}
      className={cx(css['dropdown-container'], controlClassName)}
    />
  );
};

type SingleValueCustomProps = SingleValueProps<Option> & {
  singleValueClassName?: string;
};

const SingleValue: React.FC<SingleValueCustomProps> = ({
  children,
  singleValueClassName,
  ...props
}) => {
  return (
    <components.SingleValue
      {...props}
      className={cx(css['single-select'], singleValueClassName)}
    >
      {children}
    </components.SingleValue>
  );
};

type ValueContainerCustomProps = ValueContainerProps<Option> & {
  valueContainerClassName?: string;
};

const ValueContainer: React.FC<ValueContainerCustomProps> = ({
  children,
  valueContainerClassName,
  ...props
}) => (
  <components.ValueContainer
    {...props}
    className={cx(css['value-container'], valueContainerClassName)}
  >
    {children}
  </components.ValueContainer>
);

type DropdownIndicator = IndicatorProps<Option> & {
  dropdownIndicatorClassName?: string;
};

const DropdownIndicator: React.FC<DropdownIndicator> = ({
  dropdownIndicatorClassName,
  ...props
}) => {
  return (
    <div>
      <components.DropdownIndicator
        {...props}
        className={css['dropdown-indicator-container']}
      >
        <ArrowDown
          className={cx(css['indicator-icon'], dropdownIndicatorClassName)}
        />
      </components.DropdownIndicator>
    </div>
  );
};

const IndicatorSeparator = () => {
  return <span />;
};

type OptionCustomType = OptionProps<Option> & {
  optionClassName?: string;
  optionSelectedClassName?: string;
  optionTextClassName?: string;
};

const Option: React.FC<OptionCustomType> = ({
  isSelected,
  children,
  optionClassName,
  optionSelectedClassName = '',
  optionTextClassName,
  ...props
}) => {
  return (
    <components.Option
      {...props}
      isSelected={isSelected}
      className={cx(
        css['option'],
        {
          [css['option--selected']]: isSelected,
          [optionSelectedClassName]: isSelected,
        },
        optionClassName
      )}
    >
      <div className={cx(css['option-text'], optionTextClassName)}>
        {children}
      </div>
    </components.Option>
  );
};

type MenuCustomType = MenuProps<Option> & {
  menuClassName?: string;
};

const Menu: React.FC<MenuCustomType> = ({
  menuClassName,
  children,
  ...props
}) => {
  return (
    <components.Menu
      {...props}
      className={cx(css['menu-container'], menuClassName)}
    >
      {children}
    </components.Menu>
  );
};

const Dropdown: React.FC<DropdownProps> = ({
  value,
  options,
  onChange,
  instanceId = 'dropdown',
  singleValueClassName,
  valueContainerClassName,
  controlClassName,
  dropdownIndicatorClassName,
  optionClassName,
  optionSelectedClassName,
  optionTextClassName,
  menuClassName,
}) => {
  return (
    <Select
      value={value}
      onChange={onChange}
      options={options}
      components={{
        Control: props => Control({ ...props, controlClassName }),
        SingleValue: props => SingleValue({ ...props, singleValueClassName }),
        IndicatorSeparator,
        DropdownIndicator: props =>
          DropdownIndicator({ ...props, dropdownIndicatorClassName }),
        Option: props =>
          Option({
            ...props,
            optionClassName,
            optionSelectedClassName,
            optionTextClassName,
          }),
        Menu: props => Menu({ ...props, menuClassName }),
        ValueContainer: props =>
          ValueContainer({ ...props, valueContainerClassName }),
      }}
      instanceId={instanceId}
      isClearable={false}
      isSearchable={false}
      classNamePrefix={'dropdown'}
    />
  );
};

export default Dropdown;
