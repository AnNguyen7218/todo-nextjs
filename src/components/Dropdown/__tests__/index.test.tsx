import React from 'react';
import { render, cleanup } from '@testing-library/react';
import Dropdown from '../Dropdown';

afterEach(cleanup);

describe('Dropdown Component', () => {
  test('Dropdown should match snapshot', () => {
    const options = [
      {
        value: 'value1',
        label: 'key1',
      },
      {
        value: 'value2',
        label: 'key2',
      },
    ];

    let value = options[0];
    const onChange = () => {
      value = options[1];
    };
    const { container } = render(
      <Dropdown value={value} options={options} onChange={onChange} />
    );
    expect(container).toMatchSnapshot();
  });
});
