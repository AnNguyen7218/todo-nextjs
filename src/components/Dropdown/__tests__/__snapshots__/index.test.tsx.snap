// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Component Dropdown should match snapshot 1`] = `
<div>
  <div
    class=" css-2b097c-container"
  >
    <div
      class="dropdown-container dropdown__control css-yk16xz-control"
    >
      <div
        class="value-container dropdown__value-container dropdown__value-container--has-value css-g1d714-ValueContainer"
      >
        <div
          class="single-select dropdown__single-value css-1uccc91-singleValue"
        >
          key1
        </div>
        <input
          aria-autocomplete="list"
          class="css-62g3xt-dummyInput"
          id="react-select-dropdown-input"
          readonly=""
          tabindex="0"
          value=""
        />
      </div>
      <div
        class="dropdown__indicators css-1hb7zxy-IndicatorsContainer"
      >
        <span />
        <div>
          <div
            aria-hidden="true"
            class="dropdown-indicator-container dropdown__indicator dropdown__dropdown-indicator css-tlfecz-indicatorContainer"
          >
            <ic-arrow-down.svg
              classname="indicator-icon"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
